### PR TITLE
Optimize cron

### LINKED ISSUE #3932: Optimize `Cron.next` algorithm
### What is the problem this feature would solve?

The algorithm for finding the next cron time can be optimized quite a bit. Currently, we start each iteration again from the top even if we have not advanced the day/month yet.

### What is the feature you are proposing to solve the problem?

...

### What alternatives have you considered?

_No response_

### PR DESCRIPTION
**Pull request by fubhy**
This should *massively* improve the efficiency (number of loops) for the `Cron.next` lookup. With this, I think we should also be able to support seconds & milisecond granularity.

That said, it's already possible to compose schedules with second or milisecond ticks within an active Cron interval... So I'm not sure how valuable that'd be tbh.

Closes #3932

### CODE REVIEW

**Comment by fubhy** [comment]
@IMax153 has agreed to review this.

Screenshot 2024-11-27 at 19 45 23

**Comment by tim-smart** [comment]
I think some more tests that cover off daylight savings, leap years etc. would be good.

**Comment by IMax153** [comment]
What Tim said :)

File: packages/effect/test/Cron.test.ts [hunk 1]
```diff
@@ -95,4 +132,34 @@ describe("Cron", () => {
...
+    const start = make("2024-03-31 00:00:00")
+    const sequence = Cron.sequence(parse("30 * * * *", berlin), start)
+    const next = () => sequence.next().value
+
+    deepStrictEqual(next(), make("2024-03-31 00:30:00"))
+    deepStrictEqual(next(), make("2024-03-31 01:30:00"))
+    deepStrictEqual(next(), make("2024-03-31 02:30:00"))
+    deepStrictEqual(next(), make("2024-03-31 03:30:00")) // Get's stuck here (time of transition)
+    deepStrictEqual(next(), make("2024-03-31 04:30:00")) // This is never reached
+  })
```
Comment (fubhy) [inline packages/effect/test/Cron.test.ts:164]: You are right, @tim-smart ... It does not handle transitions across daylight savings times currently :(

**Comment by fubhy** [comment]
> I think some more tests that cover off daylight savings, leap years etc. would be good.

Can you verify if this makes any sense now? I believe it does but I am not super confident ...

File: packages/effect/src/Cron.ts [hunk 2]
```diff
@@ -285,66 +338,99 @@ export const match = (cron: Cron, date: DateTime.DateTime.Input): boolean => {
...
+  const zoned = dateTime.unsafeMakeZoned(now ?? new Date(), {
+    timeZone: Option.getOrUndefined(cron.tz)
+  })
 
-  const restrictMinutes = minutes.size !== 0
-  const restrictHours = hours.size !== 0
-  const restrictDays = days.size !== 0
-  const restrictMonths = months.size !== 0
-  const restrictWeekdays = weekdays.size !== 0
+  const adjustDst = dateTime.isUtc(zoned) ? constVoid : (current: Date) => {
```
Comment (tim-smart) [inline packages/effect/src/Cron.ts:345]: .isUtc will always be false here

Comment (fubhy) [reply packages/effect/src/Cron.ts]: Gotcha. The intent here was to check if the Cron instance is set to UTC. I adapted it to cover that now.

**Review by tim-smart — APPROVED** [review]

**Pull request merged**